### PR TITLE
8336754: Remodel TypeAnnotation to "has" instead of "be" an Annotation

### PR DESCRIPTION
**Pull request by liach**
`TypeAnnotation` is not an annotation, as it should not be used in places like `AnnotationValue.ofAnnotation`. Thus it's remodeled to contain an annotation at a given location instead of to be an annotation.

Depends on #20205.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change requires CSR request [JDK-8336794](https://bugs.openjdk.org/browse/JDK-8336794) to be approved

### Issues
 * [JDK-8336754](https://bugs.openjdk.org/browse/JDK-8336754): Remodel TypeAnnotation to "has" instead of "be" an Annotation (**Bug** - P4)
 * [JDK-8336794](https://bugs.openjdk.org/browse/JDK-8336794): Remodel TypeAnnotation to "have" instead of "be" an Annotation (**CSR**)

### Reviewers
 * [Adam Sotona](https://openjdk.org/census#asotona) (@asotona - **Reviewer**)

### Contributors
 * Alex Buckley `<abuckley@openjdk.org>`

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/20247/head:pull/20247` \
`$ git checkout pull/20247`

Update a local copy of the PR: \
`$ git checkout pull/20247` \
`$ git pull https://git.openjdk.org/jdk.git pull/20247/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 20247`

View PR using the GUI difftool: \
`$ git pr show -t 20247`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/20247.diff">https://git.openjdk.org/jdk/pull/20247.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/20247#issuecomment-2237920318)

### CODE REVIEW

File: src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java [hunk 1]
```diff
@@ -283,20 +281,78 @@ private static int skipTypeAnnotation(ClassReader classReader, int p) {
...
         }
     }
 
+    private static int labelToBci(LabelContext lr, Label label, TypeAnnotation ta) {
+        //helper method to avoid NPE
+        if (lr == null) throw new IllegalArgumentException("Illegal targetType '%s' in TypeAnnotation outside of Code attribute".formatted(ta.targetInfo().targetType()));
+        return lr.labelToBci(label);
+    }
+
+    public static void writeTypeAnnotation(BufWriterImpl buf, TypeAnnotation ta) {
```
Comment (asotona) [inline src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java:302]: Is there any reason to move writeTypeAnnotation from UnboundAttribute?

Comment (liach) [reply src/java.base/share/classes/jdk/internal/classfile/impl/AnnotationReader.java:302]: This is for consistency with reading annotations: they are now defined in the same file.

**Review by asotona — APPROVED** [review]
Looks good to me.

**Comment by liach** [comment]
Unfortunately had to force push: the old commit histories now cause conflicts, so I had squashed them into one and applied onto the latest master.

**Review by asotona — APPROVED** [review]

File: test/langtools/lib/annotations/annotations/classfile/ClassfileInspector.java [hunk 2]
```diff
@@ -1197,8 +1197,8 @@ public void annoMatcher(Attribute<?> attr, ExpectedAnnotation expected) {
         switch (attr) {
             case RuntimeVisibleTypeAnnotationsAttribute rvattr -> {
                 if (expected.matchVisibility(true)) {
-                    for(Annotation anno : rvattr.annotations()) {
-                        expected.matchAnnotation(anno);
+                    for(var anno : rvattr.annotations()) {
```
Comment (turbanoff) [inline test/langtools/lib/annotations/annotations/classfile/ClassfileInspector.java:1200]:
```suggestion
                    for (var anno : rvattr.annotations()) {
```

Comment (liach) [reply test/langtools/lib/annotations/annotations/classfile/ClassfileInspector.java:1200]: There are 43 `for(` and 2 `for (` in this file, so fixing a single occurrence isn't too helpful; for consistency I think you can create a dedicated RFE for running IDE formatters over all jdk/classfile tests. I will review it if you submit such a patch.

**Comment by liach** [comment]
I have added a few more paragraphs in response to the request for comments on default values and repeatable annotations. Below are the excerpts, for `Annotation` and `AnnotationElement::name`.

```
 * In an annotation in Java source code, elements in the annotation interface
 * with default values may not be compiled into element-value pairs unless an
 * explicit value is provided. ({@jls 9.6.2}) The default value is derived from
 * the {@link AnnotationDefaultAttribute AnnotationDefault} attribute on the
 * method representing the annotation interface element in the class file
 * representing the annotation interface.
 * <p id="repeatable">
 * Multiple annotations of the same interface <i>A</i> in Java source code
 * ({@jls 9.7.5}) are represented by the {@linkplain AnnotationValue.OfAnnotation
 * annotation-valued} array elements of the {@linkplain AnnotationValue.OfArray
 * array-valued} element named {@code value} of a container annotation of the
 * containing annotation interface of <i>A</i>. ({@jls 9.6.3})
```
```
     * <p>
     * A single-element annotation ({@jls 9.7.3}) in Java source code is
     * compiled to an {@link Annotation} with exactly one {@linkplain
     * Annotation#elements key-value pair} where the element has the name
     * {@code value}.
     * <p>
     * Multiple annotations of the same interface <i>A</i> in Java source code
     * is compiled to an implicitly declared container annotation ({@jls 9.7.5})
     * with exactly one {@linkplain Annotation#elements key-value pair} where
     * the element has the name {@code value} and the type of the element is the
     * {@linkplain AnnotationValue.OfArray array} whose component type is the
     * {@linkplain AnnotationValue.OfAnnotation annotation interface <i>A</i>}.
```

**Comment by liach** [comment]
I have implemented the suggestions (compiled to -> represented by, key-value -> element value) and tried to simplify the wording. Please review the current version:
```
 * <p>
 * In an annotation in Java source code, elements in the annotation interface
 * with default values may not be represented unless an explicit value is provided.
 * ({@jls 9.6.2}) The default value is derived from the {@link AnnotationDefaultAttribute
 * AnnotationDefault} attribute on the method representing the annotation
 * interface element, in the class file representing the annotation interface.
 * <p id="repeatable">
 * Multiple annotations of the same interface <i>A</i> in Java source code
 * ({@jls 9.7.5}) are represented by the {@linkplain AnnotationValue.OfAnnotation
 * annotation-valued} array elements of the {@linkplain AnnotationValue.OfArray
 * array-valued} element named {@code value} of a container annotation, whose
 * interface is the containing annotation interface of <i>A</i>. ({@jls 9.6.3})

// location in class files to source code construct/type section
 * If this annotation is of a {@linkplain ##repeatable repeatable} annotation
 * interface <i>A</i>, is an array element of an array-valued element named
 * {@code value} in a container annotation, and the interface of the container
 * annotation is the containing annotation interface <i>AC</i> of <i>A</i>,
 * this annotation represents a base annotation of type <i>A</i>, which applies
 * to the same source code construct or type as the container annotation of
 * type <i>AC</i>.

// On AnnotationElement::name
     * @apiNote
     * In Java source code, by convention, the name of the sole element in a
     * single-element annotation interface is {@code value}. ({@jls 9.6.1})
     * A single-element annotation ({@jls 9.7.3}) declares the element value
     * for the {@code value} element. The single element of a containing
     * annotation interface that holds {@linkplain Annotation##repeatable
     * multiple} base annotations is also named {@code value}. ({@jls 9.6.3})
```

**Review by asotona — APPROVED** [review]

**Comment by liach** [comment]
Swallowed the 2 conflicts from #20513: The tags are already correct in this patch.

**Review by asotona — APPROVED** [review]

**Comment by liach** [comment]
Thanks for the reviews!

/integrate
